src/cargo/util/rustc.rs

```diff
@@ -62,4 +75,157 @@ impl Rustc {
+                let empty = CacheData {
+                    rustc_fingerprint,
+                    outputs: HashMap::new(),
+                };
+                let mut dirty = true;
+                let data = match read(&cache_location) {
+                    Ok(data) => {
+                        if data.rustc_fingerprint == rustc_fingerprint {
+                            info!("reusing existing rustc info cache");
+                            dirty = false;
+                            data
+                        } else {
+                            info!("different compiler, creating new rustc info cache");
+                            empty
+                        }
+                    }
+                    Err(e) => {
+                        info!("failed to read rustc info cache: {}", e);
+                        empty
+                    }
+                };
+                return Cache {
+                    cache_location: Some(cache_location),
+                    dirty,
+                    data,
+                };
+
+                fn read(path: &Path) -> CargoResult<CacheData> {
+                    let json = paths::read(path)?;
+                    Ok(serde_json::from_str(&json)?)
+                }
+            }
+            (_, fingerprint) => {
+                if let Err(e) = fingerprint {
+                    warn!("failed to calculate rustc fingerprint: {}", e);
+                }
+                info!("rustc info cache disabled");
+                Cache {
+                    cache_location: None,
+                    dirty: false,
+                    data: CacheData::default(),
+                }
+            }
+        }
+    }
+
+    fn cached_output(&mut self, cmd: &ProcessBuilder) -> CargoResult<(String, String)> {
+        let calculate = || {
+            let output = cmd.exec_with_output()?;
+            let stdout = String::from_utf8(output.stdout)
+                .map_err(|_| internal("rustc didn't return utf8 output"))?;
+            let stderr = String::from_utf8(output.stderr)
+                .map_err(|_| internal("rustc didn't return utf8 output"))?;
+            Ok((stdout, stderr))
+        };
+        if self.cache_location.is_none() {
+            info!("rustc info uncached");
+            return calculate();
+        }
+
+        let key = process_fingerprint(cmd);
+        match self.data.outputs.entry(key) {
+            Entry::Occupied(entry) => {
+                info!("rustc info cache hit");
+                Ok(entry.get().clone())
+            }
+            Entry::Vacant(entry) => {
+                info!("rustc info cache miss");
+                let output = calculate()?;
+                entry.insert(output.clone());
+                self.dirty = true;
+                Ok(output)
+            }
+        }
+    }
+}
+
+impl Drop for Cache {
+    fn drop(&mut self) {
+        match (&self.cache_location, self.dirty) {
```

Comment:
Perhaps this could be started with:

```rust
if !self.dirty {
    return 
}
```

Reply:
Er well I'll leave it up to you, I'm just personally always looking to aggressively reduce indent!

Reply:
Yeah, it looks much nicer that way, because I don't have to wrangle with not moving out cahce_location as well!